File: C++/Navio+/RCInput_Navio.cpp
#include "RCInput_Navio.h"
#include "RCOutput_Navio.h"

RCInput_Navio::RCInput_Navio()
{

}

RCInput_Navio::~RCInput_Navio()
{
}

void RCInput_Navio::ppmOnEdgeTrampolin(int gpio, int level, uint32_t tick, void *userdata)
{
    RCInput_Navio *self = (RCInput_Navio *) userdata;

    self->ppmOnEdge(gpio, level, tick);
}

void RCInput_Navio::ppmOnEdge(int gpio, int level, uint32_t tick)
    {
    if (level == 0) {
        deltaTime = tick - previousTick;
        previousTick = tick;

        if (deltaTime >= ppmSyncLength) { // Sync
            currentChannel = 0;

            // RC output
            for (int i = 0; i < ppmChannelsNumber; i++)
                pwm->set_duty_cycle(i+3, channels[i] / 1000); // 1st Navio RC output is 3

        }
        else
            if (currentChannel < ppmChannelsNumber)
                channels[currentChannel++] = deltaTime;
    }
}

void RCInput_Navio::initialize()
{
    Pin pin(outputEnablePin);

    if (pin.init()) {
        pin.setMode(Pin::GpioModeOutput);
        pin.write(0); /* drive Output Enable low */
    } else {
        fprintf(stderr, "Output Enable not set. Are you root?");
    }

    // Servo controller setup

    pwm = new RCOutput_Navio();
    pwm->initialize();
    pwm->setFrequency(servoFrequency);

    // GPIO setup

    gpioCfgClock(samplingRate, PI_DEFAULT_CLK_PERIPHERAL, 0);
    gpioInitialise();
    gpioSetMode(4,PI_INPUT);

    previousTick = gpioTick();
    gpioSetAlertFuncEx(ppmInputGpio, RCInput_Navio::ppmOnEdgeTrampolin, this);
}

int RCInput_Navio::read(int ch)
{
    return channels[ch];
}

File: Python/navio/Navio/RCInput_Navio.py
import RPi.GPIO as GPIO
import pigpio

class RCInput_Navio():

    CHANNEL_COUNT = 8
    channels = [0,0,0,0,0,0,0,0]

    def cbf(self, gpio, level, tick):
        if (level == 0):
            deltaTime = tick - self.previousTick
            self.previousTick = tick

            if (deltaTime >= 4000):
                self.currentchannel = 0
            else:
                if (self.currentchannel < self.CHANNEL_COUNT):
                    self.channels[self.currentchannel] = deltaTime
                    self.currentchannel += 1

    def __init__(self):

        GPIO.setmode(GPIO.BCM)
        GPIO.setup(27, GPIO.OUT)
        GPIO.output(27, GPIO.LOW)

        self.pi = pigpio.pi()
        self.pi.set_mode(4, pigpio.INPUT)

        self.currentchannel = 0
        self.previousTick = self.pi.get_current_tick()
        self.cb1 = self.pi.callback(4, pigpio.EITHER_EDGE, self.cbf)

    def read(self, ch):

        return self.channels[ch]

